cryptoprotocol/src/main/java/org/cryptimeleon/incentivesystem/cryptoprotocol/proof/SpendDeductZkp.java

```diff
@@ -0,0 +1,164 @@
+        var dsrndStar1Var = builder.addZnVariable("dsrndStar1", zn);
+        var vVar = builder.addZnVariable("v", zn);
+        var zVar = builder.addZnVariable("z", zn);
+        var tVar = builder.addZnVariable("t", zn);
+        var zStarVar = builder.addZnVariable("zStar", zn);
+        var tStarVar = builder.addZnVariable("tStar", zn);
+        var uStarVar = builder.addZnVariable("uStar", zn);
+        var uStarInverseVar = builder.addZnVariable("uStarInverse", zn);
+        var eskDecVarVector = new SchnorrZnVariable[pp.getNumEskDigits()];
+        var rVector = new SchnorrZnVariable[pp.getNumEskDigits()];
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            eskDecVarVector[i] = builder.addZnVariable("eskStarUserDec_" + i, zn);
+            rVector[i] = builder.addZnVariable("r_" + i, zn);
+        }
+
+        // c0=usk*gamma+dsrnd0
+        var c0statement = new ExponentSumExpr(new ExponentMulExpr(uskVar, new ExponentConstantExpr(commonInput.gamma)), dsrnd0Var);
+        builder.addSubprotocol("c0=usk*gamma+dsrnd0", new LinearExponentStatementFragment(c0statement.isEqualTo(commonInput.c0), zn));
+
+        // c1=esk*gamma+dsrnd1
+        var c1statement = new ExponentSumExpr(new ExponentMulExpr(eskVar, new ExponentConstantExpr(commonInput.gamma)), dsrnd1Var);
+        builder.addSubprotocol("c1=esk*gamma+dsrnd1", new LinearExponentStatementFragment(c1statement.isEqualTo(commonInput.c1), zn));
+
+        // dsid=w^esk
+        var dsidEskStatement = w.pow(eskVar).isEqualTo(commonInput.dsid);
+        builder.addSubprotocol("dsid=w^esk", new LinearStatementFragment(dsidEskStatement));
+
+        // C=(H.pow(usk, esk, dsrnd_0, dsrnd_1, v, z, t),g_1) split into two subprotocols
+        var commitmentC0Statement = H.pow(new Vector<>(
+                uskVar,
+                eskVar,
+                dsrnd0Var,
+                dsrnd1Var,
+                vVar,
+                zVar,
+                tVar
+        )).reduce(GroupElementExpression::op).isEqualTo(commonInput.commitmentC0);
+        builder.addSubprotocol("C0", new LinearStatementFragment(commitmentC0Statement));
+        builder.addSubprotocol("C1", new LinearStatementFragment(commonInput.commitmentC1.expr().isEqualTo(pp.getG1())));
+
+        // C=(H.pow(usk, \sum_i=0^k[esk^*_(usr,i) * base^i], dsrnd^*_0, dsrnd^*_1, v-k, z^*, t^*), g_1^(u^*)) split into two subprotocols
+        // We use the sum to combine the esk^*_usr = \sum proof with the C=.. proof
+        var exponents = new Vector<>(uskVar, BigInteger.ZERO, dsrndStar0Var, dsrndStar1Var, vVar.sub(zn.valueOf(commonInput.k)), zStarVar, tStarVar); // 0 because h_2^(esk^*_usr) is handled separately
+        GroupElementExpression eskPart = groupG1.getNeutralElement().expr();
+        for (int i = 0; i < pp.getNumEskDigits(); i++) {
+            eskPart = eskPart.op(H.get(1).pow(eskDecVarVector[i].mul(pp.getEskDecBase().pow(BigInteger.valueOf(i)))));
```

Comment:
This way of writing this causes numDigits group operations and exponentiations. It can be rewritten as `H.get(1).pow(vectorOfEskStar.innerProduct(vectorOfPowersOfTwo))`, i.e. a single exponentiation.

Reply:
Even further, there's nothing stopping you from just replacing BigInteger.ZERO above with `vectorOfEskStar.innerProduct(vectorOfPowersOfTwo)`.

Reply:
Is there some `SchnorrZnVariableVector` or so that supports this? Currently, the `eskStarVector` is a `SchnorrZnVariable[]`.

Reply:
There is an ExponentExpressionVector that fits this need.

Reply:
Added this, also for rVector